syntax = "proto3";

option java_multiple_files = true;
option java_package = "dmlc.federated.xgboost";
option java_outer_classname = "XGBProto";
option objc_class_prefix = "XGB";

package xgbcomm;

service XgbService {
  rpc GetEncriptedGradPairs_ (stream GradPairsRequest) returns (stream GradPairsResponse) {}
  rpc GetEncriptedSplits_ (stream SplitsRequest) returns (stream SplitsResponse) {}
  rpc GetEncriptedGradPairs (GradPairsRequest) returns (GradPairsResponse) {}
  rpc GetEncriptedSplits (SplitsRequest) returns (SplitsResponse) {}
}

message GradPairsRequest {
  uint32 version = 1;
}

message GradPairsResponse {
  uint32 version = 1;
  repeated MpzType encripted_grad_pairs = 2;
}

message SplitsRequest {
  uint32 version = 1;
}

message SplitsResponse {
  uint32 version = 1;
  repeated EncriptedSplit encripted_splits = 2;
}

message EncriptedSplit {
  string mask_id = 1;
  MpzType encripted_grad_pair_sum = 2;
}

message MpzType {
  /* Number of *limbs* allocated and pointed to by the _mp_d field.  */
  int32 _mp_alloc = 1;
  /* abs(_mp_size) is the number of limbs the last field points to.
     If _mp_size is negative this is a negative number.  */
  int32 _mp_size = 2;
  /* Pointer to the limbs.  */
  repeated uint64 _mp_d = 3;
}
